dhcpv4/ztp/ztp.go

```diff
@@ -0,0 +1,139 @@
+package ztpv4
+
+import (
+	"bytes"
+	"errors"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/insomniacslk/dhcp/dhcpv4"
+)
+
+// VendorData is optional data a particular vendor may or may not include
+// in the Vendor Class options.  All values are optional and will be zero
+// values if not found.
+type VendorData struct {
+	VendorName string
+	Model      string
+	Serial     string
+}
+
+var errVendorOptionMalformed = errors.New("malformed vendor option")
+
+// VendorDataV4 will try to parse dhcp4 options data looking for more specific
+// vendor data (like model, serial number, etc).  If the options are missing
+func VendorDataV4(packet *dhcpv4.DHCPv4) VendorData {
+	vd := VendorData{}
+
+	if err := parseV4VendorClass(&vd, packet); err != nil {
+		glog.Errorf("failed to parse vendor data from vendor class: %v", err)
+	}
+
+	if err := parseV4VIVC(&vd, packet); err != nil {
+		glog.Errorf("failed to parse vendor data from vendor-idenitifying vendor class: %v", err)
+	}
+
+	return vd
+}
+
+// parseV4Opt60 will attempt to look at the Vendor Class option (Option 60) on
+// DHCPv4.  The option is formatted as a string with the content being specific
+// for the vendor, usually using a deliminator to separate the values.
+// See: https://tools.ietf.org/html/rfc1533#section-9.11
+func parseV4VendorClass(vd *VendorData, packet *dhcpv4.DHCPv4) error {
+	opt := packet.GetOneOption(dhcpv4.OptionClassIdentifier)
+	if opt == nil {
+		return nil
+	}
+	vc := opt.(*dhcpv4.OptClassIdentifier).Identifier
+
+	switch {
+	// Arista;DCS-7050S-64;01.23;JPE12221671
+	case strings.HasPrefix(vc, "Arista;"):
+		p := strings.Split(vc, ";")
+		if len(p) < 4 {
+			return errVendorOptionMalformed
+		}
+
+		vd.VendorName = p[0]
+		vd.Model = p[1]
+		vd.Serial = p[3]
+		return nil
+
+	// ZPESystems:NSC:002251623
+	case strings.HasPrefix(vc, "ZPESystems:"):
+		p := strings.Split(vc, ":")
+		if len(p) < 3 {
+			return errVendorOptionMalformed
+		}
+
+		vd.VendorName = p[0]
+		vd.Model = p[1]
+		vd.Serial = p[2]
+		return nil
+
+	// Juniper option 60 parsing is a bit more nuanced.  The following are all
+	// "valid" indetifing stings for Juniper:
+	//    Juniper-ptx1000-DD576      <vendor>-<model>-<serial
+	//    Juniper-qfx10008           <vendor>-<model> (serial in hostname option)
+	//    Juniper-qfx10002-361-DN817 <vendor>-<model>-<serial> (model has a dash in it!)
+	case strings.HasPrefix(vc, "Juniper-"):
+		// strip of the prefix
+		vc := vc[len("Juniper-"):]
```

Comment:
why not splitting here too? It seems that all the fields are dash-separated in every case, and it would simplify the following code

Reply:
I think this was done this way because the model may have a dash in it, example: Juniper-qfx10002-361-DN817.

Reply:
I see - what I meant is, why don't we just split by dashes, and use the number of fields to know if the model has a dash in it or not? E.g. if len is 4, the two middle fields are the parts of the model name. This means you have to rejoin them, but it simplifies all the rest of the code block